tiledb/sm/group/group_details.cc

```diff
@@ -0,0 +1,280 @@
+}
+
+Status GroupDetails::clear() {
+  members_.clear();
+  members_by_name_.clear();
+  members_vec_.clear();
+  members_to_modify_.clear();
+
+  return Status::Ok();
+}
+
+void GroupDetails::add_member(const tdb_shared_ptr<GroupMember>& group_member) {
+  std::lock_guard<std::mutex> lck(mtx_);
+  const std::string& uri = group_member->uri().to_string();
+  members_.emplace(uri, group_member);
+  members_vec_.emplace_back(group_member);
+  if (group_member->name().has_value()) {
+    members_by_name_.emplace(group_member->name().value(), group_member);
+  }
+}
+
+void GroupDetails::delete_member(
+    const tdb_shared_ptr<GroupMember>& group_member) {
+  std::lock_guard<std::mutex> lck(mtx_);
+  const std::string& uri = group_member->uri().to_string();
+  auto it = members_.find(uri);
+  if (it != members_.end()) {
+    auto name = it->second->name();
+    members_.erase(it);
+    if (group_member->name().has_value()) {
+      members_by_name_.erase(group_member->name().value());
+    } else if (name.has_value()) {
+      members_by_name_.erase(name.value());
+    }
+    for (size_t i = 0; i < members_vec_.size(); i++) {
+      if (members_vec_[i] == it->second) {
+        members_vec_.erase(members_vec_.begin() + i);
+        break;
+      }
+    }
+  }
+}
+
+Status GroupDetails::mark_member_for_addition(
+    const URI& group_member_uri,
+    const bool& relative,
+    std::optional<std::string>& name,
+    StorageManager* storage_manager) {
+  std::lock_guard<std::mutex> lck(mtx_);
+  // TODO: Safety checks for not double adding, making sure its remove + add,
+  // etc
+
+  URI absolute_group_member_uri = group_member_uri;
+  if (relative) {
+    absolute_group_member_uri =
+        group_uri_.join_path(group_member_uri.to_string());
+  }
+  ObjectType type = ObjectType::INVALID;
+  RETURN_NOT_OK(storage_manager->object_type(absolute_group_member_uri, &type));
+  if (type == ObjectType::INVALID) {
+    return Status_GroupError(
+        "Cannot add group member " + absolute_group_member_uri.to_string() +
+        ", type is INVALID. The member likely does not exist.");
+  }
+
+  auto group_member = tdb::make_shared<GroupMemberV2>(
+      HERE(), group_member_uri, type, relative, name, false);
+
+  members_to_modify_.emplace_back(group_member);
+
+  return Status::Ok();
+}
+
+Status GroupDetails::mark_member_for_removal(const URI& uri) {
+  return mark_member_for_removal(uri.to_string());
+}
+
+Status GroupDetails::mark_member_for_removal(const std::string& uri) {
+  std::lock_guard<std::mutex> lck(mtx_);
+
+  auto it = members_.find(uri);
+  auto it_name = members_by_name_.find(uri);
+  if (it != members_.end()) {
+    auto member_to_delete = make_shared<GroupMemberV2>(
+        it->second->uri(),
+        it->second->type(),
+        it->second->relative(),
+        it->second->name(),
+        true);
+    members_to_modify_.emplace_back(member_to_delete);
+    return Status::Ok();
+  } else if (it_name != members_by_name_.end()) {
```

Comment:
Not a fan of overloading the input parameter here. At least call it `uri_or_name` clarity...